### PR TITLE
YOYO Application - Kurt Ma. Coll

### PR DESCRIPTION
**Pull request by makoru-hikage**
#### Q: What design patterns did you use inside the application?

A: I used the Builder pattern to to do method chaining, it is easy when there are joined tables or multiple steps and SQL queries.

I used the repository pattern for easier database operations.

#### Q: What libraries did you use?
A: Here are the libraries I have used (except from core libraries):
* Dancer2;
	* Quite simple to use. It is like PHP's [Slim framework](https://www.slimframework.com/); Bare minimum but out-of-the-box and sufficient for my needs. The documentation is ample.
	* Catalyst is too heavy and Mojolicious seems heavier than Dancer2.
* Dancer2::Plugin::DBIC;
	* Just for the synergy of Dancer2 and DBIx
* Dancer2::Plugin::Passphrase;
	* Just for the sake of password hashes, nothing else.
* Validation::Class;
	* The simplest validator I found. Although I can't make it validate the uniqueness of a username and similar values.
* DBIx::Class::TimeStamp;
	* Still can't make it work. It is just there to change the "updated_at" timestamp. 
* DBIx::Class;
	* Gave me a headache. It is my ORM and quite easy to use, until you use the advanced features. Somehow managed to use its advanced features. The documentation is so ample.

#### Q: What major challenges did you encounter when making the app?

A: Some hours of minimal burnout. The documentation is ample in DBIx and Dancer2, but some features are not clear enough to me... due to the Perl syntax they utilize, such as this:

`[ key1 => "value1", key2 => "value2" ]`

Confused me because it is an anonymous function with "fat comma" until I see the [examples from Dancer2](http://search.cpan.org/~cromedome/Dancer2-0.205000/lib/Dancer2/Manual.pod#Declaring_Routes)(see how the route declaration is written). That anonymous array has two elements: an anonymous hash and another anonymous hash. Each hash has only one element.

Also the time I spent (almost 30 days) is spent on learning Perl and it's advanced syntax then doing this project. Prior to the first email sent to me by Yoyo, this is the [summary of my Perl skills](https://gist.github.com/makoru-hikage/e896e1a78bb58686741e7c1e23b86177).

I made the script just for generating PHP classes. (XAMPP has Perl in it, so... that's it.) After I made those script, I haven't written in Perl for about 2 years.

I like Perl for extreme flexibility.

#### Q: If you had more time, what additional features would you like to add?

A: The GUI and API resources for User entities.

### CODE REVIEW

File: yoyotest/INSTALL.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Installation
+
+## Windows:
+
+1. Install Strawberry Perl (It comes with CPAN)
+2. Do `git clone https://github.com/makoru-hikage/api-org-app`
+3. Assure that you have `Carton`, if you don't have, just do `cpan Carton`, it's your project package manager (like PHP's Composer)
+4. Traverse to the `api-org-app` folder (which resulted from git clone) then do `carton install`
+5. Once the installation finishes, run Plack by doing `plackup -p 3000` when you are in the `bin` folder. Everybody knows `bin`
```
Comment (tedmdelacruz) [inline yoyotest/INSTALL.md]: I think it should be `carton exec plackup -p 3000`

File: yoyotest/cpanfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+requires "Dancer2" => "0.205000";
+requires "Dancer2::Plugin::DBIC" => "0.0013";
+requires "Dancer2::Plugin::Passphrase" => "3.3.0";
+requires "Validation::Class" => "7.900057";
+requires "DBIx::Class::TimeStamp" => "0.14";
+requires "DBIx::Class" => "0.082840";
```
Comment (tedmdelacruz) [inline yoyotest/cpanfile]: Missing dependencies: ```
requires "Dancer2::Session::Cookie" => "0.0006";
requires "Plack::Middleware::Deflater" => "0.12"
```

**Comment by tedmdelacruz** [comment]
Please rename PR with your name, something like `YOYO Application - <your name>`

File: yoyotest/public/css/win98.css [hunk 3]
```diff
@@ -0,0 +1,210 @@
+.block-centered {
```
Comment (tedmdelacruz) [inline yoyotest/public/css/win98.css]: Please don't just copy-paste open-source code without attribution if you're not using package  managers such as `bower` or `npm`

http://contra.io/windows_98.css/

File: yoyotest/INSTALL.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+6. ????
+7. YEHEY!
+
+## Linux:
+1. Most linux distros have perl installed on them, if there is no perl, read the docs of your distro's package manager to install perl (if you are using Gentoo or LFS, then compile it yourself)
+2. Again, download Carton by using cpan, if you dont have it...
+3. ... AHHHHHHHHH to continue this list, see items 4-7 in Windows section.
+
+ASSURE THAT YOU HAVE MYSQL AND IMPORT THE `Yoyotest.sql` INTO IT, THEN KEEP IT RUNNING.
```
Comment (tedmdelacruz) [inline yoyotest/INSTALL.md]: Also put a note that the database configuration should be set in `yoyotest/environments/development.yml`

* Also a minor note, database username and passwords shouldn't be set in version control

**Comment by tedmdelacruz** [comment]
Doesn't seem to work on my end

File: yoyotest/public/css/win98.css (hunk 3 above)
Comment (makoru-hikage) [inline yoyotest/public/css/win98.css]: Sorry

Comment (makoru-hikage) [inline yoyotest/public/css/win98.css]: Here: https://github.com/makoru-hikage/api-org-app/commit/07e4ac993b7550f13a2aea9189a28e8ab3d9945f

File: yoyotest/cpanfile (hunk 2 above)
Comment (makoru-hikage) [inline yoyotest/cpanfile]: Fixed it: https://github.com/makoru-hikage/api-org-app/commit/4b992e92f13c2d0e2e53e100c9d9a58fb74c7039

File: yoyotest/INSTALL.md (hunk 4 above)
Comment (makoru-hikage) [inline yoyotest/INSTALL.md]: Removed user and pass: https://github.com/makoru-hikage/api-org-app/commit/de83916cc60059a02b753e2465de4b59b6b937a1

Comment (makoru-hikage) [inline yoyotest/INSTALL.md]: I have set my `yoyotest/environments/development.yml`; what kind of database configuration?

**Comment by makoru-hikage** [comment]
My apologies if GUI is not yet, functional, Sir @tedmdelacruz. Would you like me to finish it?

**Comment by makoru-hikage** [comment]
I'll finish the GUI.

File: yoyotest/INSTALL.md (hunk 4 above)
Comment (tedmdelacruz) [inline yoyotest/INSTALL.md]: I mean just add a note in the setup instructions (`INSTALL.md`) on how to setup the database, which involves editing `yoyotest/environments/development.yml`

**Comment by tedmdelacruz** [comment]
> My apologies if GUI is not yet, functional, Sir @tedmdelacruz. Would you like me to finish it?

@makoru-hikage We just need to get it working as instructed in the exam

**Comment by makoru-hikage** [comment]
Alright, Sir.

**Comment by tedmdelacruz** [comment]
@makoru-hikage Also, please don't address us with _sir_ 😛

**Comment by jcyu0208** [comment]
@makoru-hikage 

Thank you for your submission. 
We'll get back to you with the results of our evaluation.

**Comment by htgc** [comment]
@makoru-hikage Sorry for the late reply. May I ask some questions?

- How long did you take to make this app?
- What programing language have you experience other than Perl, PHP and JavaScript?
- Which OS do you usually use either Windows or Linux or Mac?

**Comment by makoru-hikage** [comment]
- Q: How long did you take to make this app?

A: The backend took me a span of about 26 days including the enhancement of my Perl skills.
The frontend took me a span of about 12 days.

- Q: What programing language have you experience other than Perl, PHP and JavaScript?

A: I had experience with C. I used GTK+ and [Gnome-DB](http://www.gnome-db.org/). The Gnome DB GUI has some significant bugs. I got two of them; the problem is that it has only 1 maintainer: Vivien Malerba, that time. Fixes took some days so I discontinued doing the app and rewrote it in PHP. However, the two bugs were fixed, nonetheless, I discontinued using C.

My first programming language was VB6, I was a 3rd year highschool student that time. My big brother's Grade System source code was the source of examples that I use, such as simple form load, show, hide, command buttons, etc. I was making quizzes back then.

Java... I made a simple grades database in Java (which was back at 2012). It is a small project just to demonstrate the skills of students on how to connect a database to a desktop application. There were groups of two. Only two of the groups used Java Netbeans, that 30-ish classmate of mine and I. (Each group has two people whilst my group has only one, due to my hubris that I discarded months ago.)

C++ because TURBO C! YAY! Ah, just the standard Programming 1 class. I'm not yet intimate with that language; My resumé does not speak about C++.

Q: Which OS do you usually use either Windows or Linux or Mac?

These recent days, I'm using Windows for it is the OS of our home computer. However, back in 2011-2014, I frequently use Linux on my [Eee PC 901](https://www.asus.com/Laptops/Eee_PC_901/) with no secondary memory but an 8GB flash drive. Back in 2011, I use Ubuntu, but it was laggy due to [Unity](https://unity.ubuntu.com/). In 2012, I settled with minimal [UberStudent](http://uberstudent.com/) because of its speed. In 2013, I used [BackBox](https://backbox.org/) because it is a hacking distro (though I've not done any actual pentesting). In 2014, I learned how to use [Arch Linux](https://www.archlinux.org/). Initially a headache, I have grown accustomed to it. It boots under a minute. I use [MATE](https://mate-desktop.org/) window manager, and [Light DM](https://wiki.archlinux.org/index.php/LightDM). Arch linux is the best I ever have. I still have it on my 16GB USB flash drive, installed in an EXT4 filesystem. (It was previously a F2FS file system.)

I've never owned a Mac PC.

**Comment by makoru-hikage** [comment]
The GUI is done. The API is done.

**Pull request closed without merging**